File: snippets/Basics/go/fibonacci.go
// Metadata Header (MANDATORY)
// -----------------------------
// Program Title: Fibonacci Sequence (Iterative)
// Author: ghostrider2606
// Date: 2025-10-11
//
// Description: Generates the Fibonacci sequence up to the n-th term using an iterative approach.
//
// Language: Go
//
// Time Complexity: O(n)
// Space Complexity: O(1)
// -----------------------------

package main

import "fmt"

// fibonacciIterative generates the first n Fibonacci numbers.
// Uses uint64 for large numbers.
func fibonacciIterative(n int) ([]uint64, error) {
    if n < 0 {
        return nil, fmt.Errorf("input must be a non-negative integer")
    }
    if n == 0 {
        return []uint64{}, nil
    }
    if n == 1 {
        return []uint64{0}, nil
    }

    sequence := make([]uint64, n)
    sequence[0] = 0
    sequence[1] = 1

    var a, b uint64 = 0, 1

    for i := 2; i < n; i++ {
        nextVal := a + b
        sequence[i] = nextVal
        
        // Shift values
        a = b
        b = nextVal
    }
    return sequence, nil
}

func main() {
    n_terms := 20
    result, err := fibonacciIterative(n_terms)
    
    if err != nil {
        fmt.Println("Error:", err)
    } else {
        fmt.Printf("Fibonacci sequence (first %d terms): %v\n", n_terms, result)
    }
}

File: snippets/Basics/java/fibonacci.java
// Metadata Header (MANDATORY)
// -----------------------------
// Program Title: Fibonacci Sequence (Iterative)
// Author: IamBisrutPyne
// Date: 2025-10-11
//
// Description: Generates the Fibonacci sequence up to the n-th term using an iterative approach.
//
// Language: Java
//
// Time Complexity: O(n)
// Space Complexity: O(1)
// -----------------------------

import java.util.ArrayList;
import java.util.List;

public class FibonacciIterative {

    /**
     * Generates the Fibonacci sequence up to the n-th term iteratively.
     * Uses long to prevent integer overflow for larger sequences.
     *
     * @param n The number of terms to generate.
     * @return A list of the first n Fibonacci numbers.
     */
    public static List<Long> generate(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("Input must be non-negative.");
        }
        
        List<Long> sequence = new ArrayList<>();
        if (n == 0) return sequence;
        
        long a = 0; // F(0)
        long b = 1; // F(1)

        sequence.add(a);
        if (n == 1) return sequence;

        sequence.add(b);
        
        for (int i = 2; i < n; i++) {
            long nextVal = a + b;
            sequence.add(nextVal);
            // Shift values
            a = b;
            b = nextVal;
        }
        return sequence;
    }

    public static void main(String[] args) {
        int n_terms = 20;
        System.out.println("Fibonacci sequence (first " + n_terms + " terms): " + generate(n_terms));
    }
}
